### ISSUE TITLE
[Bug]: DSA sparse_attention_backward (SM90/SM100): NaN gradients on invalid/empty/all-sink top-k rows, GPU deadlock on empty rows, wrong d_sink, and accuracy loss under large logits

### ISSUE THREAD

**Issue by wangfanChris** [NVIDIA/cudnn-frontend#676]
### Before submitting

- [x] #677

### Affected area

None

### Description

The DeepSeek Sparse Attention backward kernel
(`python/cudnn/deepseek_sparse_attention/sparse_attention_backward/dsa_bwd_sm90.py`,
exposed through `cudnn.DSA.sparse_attention_backward_wrapper`) is numerically
unsafe for several categories of input that are valid and common in practice.
The failures are deterministic and reproduce on a single GPU. Any training run
that hits a padded/empty/saturating top-k row gets NaN gradients (or a hung GPU),
which silently corrupts or stalls the whole step.

Five distinct, independently-reproducible symptoms:

1. **NaN gradients for invalid / padded top-k rows.** When a `topk_idxs` row
   contains entries outside `[0, s_kv)` — the `-1` padding used when the active
   key count is below the allocated `topk` width, out-of-range positive indices,
   or entries past `topk_length` — those lanes are not masked out of the softmax.
   The kernel dereferences the raw index and feeds a garbage score into
   `exp2(score − lse)`, which overflows to `inf` and then `inf·dP = NaN` in `dq`.

2. **GPU deadlock on fully-empty rows.** A query row with `topk_length = 0` and
   the sink disabled (`attn_sink = -inf`) never returns: `torch.cuda.synchronize()`
   hangs indefinitely (>150 s, 100% GPU utilization). The empty-row CTA does not
   reach the barrier that the rest of the launch waits on.

3. **All-NaN `dq` when the sink owns all probability mass.** With
   `attn_sink → +inf` the KV probability is 0 for every key. The log-sum-exp with
   sink computes `+inf − +inf = NaN` (and the delta path computes `0·inf`), so
   every head of `dq` becomes NaN.

4. **Wrong `d_sink`.** The softmax delta `Σ p·dP` is read from the cached BF16
   forward output as `O·dO` instead of being recomputed from `Q·K·dO`. After BF16
   rounding of `O`, the delta that feeds `d_sink` is off by a term that only
   cancels in full precision, so `d_sink` is wrong for **all** heads (observed max
   abs error 9.77e-4 against an FP64 reference; tolerance 2e-5).

5. **Accuracy collapse under large or shifted logits.** Adding a constant to all
   logits (mathematically gradient-invariant), a saturated softmax (one dominant
   key), or a high activation scale all drive `dq`/`dkv` to NaN or to large
   relative error (up to `relative_l2 ≈ 168`, i.e. 16800%). The single-pass online
   softmax statistics accumulate catastrophic cancellation when the block-local
   maximum differs from the global maximum.

Both the SM90 and SM100 code paths are affected. Only the SM90 path was executed
here because the test machine's H200 reports `compute capability 9.0`; the SM100
tests skip automatically. The SM100 kernel (`dsa_bwd_sm100.py`) shares the same
delta-from-`O·dO`, sink, and single-pass-statistics structure and is expected to
reproduce items 1–5 identically.

### Steps to reproduce

Minimal standalone repro (each block fails on its own):

```python
import math, torch
from cudnn import DSA

device = torch.device("cuda")

# --- Symptom 1: invalid (padded) top-k rows -> NaN dq -----------------------
s_q, s_kv, num_heads = 1, 1, 128
head_dim, head_dim_v = 576, 512
softmax_scale = 192**-0.5

q    = torch.randn(s_q, num_heads, head_dim,   dtype=torch.bfloat16, device=device) * 8
kv   = torch.randn(s_kv, head_dim,             dtype=torch.bfloat16, device=device) * 8
dout = torch.randn(s_q, num_heads, head_dim_v, dtype=torch.bfloat16, device=device)
attn_sink = torch.full((num_heads,), float("-inf"), dtype=torch.float32, device=device)

topk = 64
topk_idxs = torch.full((s_q, topk), -1, dtype=torch.int32, device=device)  # all padding
topk_idxs[:, 0] = 0                                                         # one valid key

out = kv[0, :head_dim_v].view(1, 1, head_dim_v).expand(s_q, num_heads, head_dim_v).contiguous()
lse = ((q.float() * kv[0].float()).sum(dim=-1) * softmax_scale).contiguous()

r = DSA.sparse_attention_backward_wrapper(
    q, kv, out, dout, lse, attn_sink, topk_idxs, softmax_scale=softmax_scale)
torch.cuda.synchronize()
assert torch.isfinite(r["dq"]).all()          # FAILS: dq is NaN


# --- Symptom 3: +inf sink -> all-NaN dq ------------------------------------
q  = torch.randn(1, num_heads, head_dim, dtype=torch.bfloat16, device=device)
kv = torch.randn(1, head_dim,            dtype=torch.bfloat16, device=device)
out  = torch.zeros(1, num_heads, head_dim_v, dtype=torch.bfloat16, device=device)
dout = torch.randn_like(out)
lse  = ((q.float() * kv[0].float()).sum(dim=-1) * softmax_scale).contiguous()
attn_sink = torch.full((num_heads,), float("inf"), dtype=torch.float32, device=device)
topk_idxs = torch.zeros(1, 1, dtype=torch.int32, device=device)

r = DSA.sparse_attention_backward_wrapper(
    q, kv, out, dout, lse, attn_sink, topk_idxs, softmax_scale=softmax_scale)
torch.cuda.synchronize()
assert torch.isfinite(r["dq"]).all()          # FAILS: dq is all NaN


# --- Symptom 2: empty row -> GPU deadlock (uncomment to observe the hang) ---
# topk_idxs   = torch.full((1, topk), torch.iinfo(torch.int32).max, dtype=torch.int32, device=device)
# topk_length = torch.zeros(1, dtype=torch.int32, device=device)
# lse = torch.full((1, num_heads), float("-inf"), dtype=torch.float32, device=device)
# out = torch.zeros(1, num_heads, head_dim_v, dtype=torch.bfloat16, device=device)
# r = DSA.sparse_attention_backward_wrapper(
#     q, kv, out, torch.randn_like(out), lse, attn_sink, topk_idxs,
#     topk_length=topk_length, softmax_scale=softmax_scale)
# torch.cuda.synchronize()                     # never returns
```

The accuracy symptoms (4, 5) are checked against an FP64 autograd reference that
computes only the *selected* KV columns, so it is contraction-shape invariant:

```python
def sparse_attention_forward_fp64(q, kv, topk_idxs, softmax_scale, attn_sink=None):
    outs, lses = [], []
    for row in range(q.shape[0]):
        sel = topk_idxs[row][topk_idxs[row] >= 0].to(torch.int64)     # drop padding
        k   = kv.index_select(0, sel)
        logits = torch.einsum("hd,kd->hk", q[row], k) * softmax_scale
        lse    = torch.logsumexp(logits, dim=-1)
        if attn_sink is None:
            p = torch.softmax(logits, dim=-1)
        else:
            p = torch.exp(logits - torch.logaddexp(lse, attn_sink.to(lse.dtype)).unsqueeze(-1))
        outs.append(torch.einsum("hk,kd->hd", p, k[:, :512]))
        lses.append(lse)
    return torch.stack(outs), torch.stack(lses)

# d_sink recomputed-delta bug (symptom 4)
q_ref  = q.to(torch.float64).requires_grad_(True)
kv_ref = kv.to(torch.float64).requires_grad_(True)
sink_ref = attn_sink.to(torch.float64).requires_grad_(True)
out_ref, lse_ref = sparse_attention_forward_fp64(q_ref, kv_ref, topk_idxs, softmax_scale, sink_ref)
r = DSA.sparse_attention_backward_wrapper(
    q, kv, out_ref.detach().to(q.dtype), dout, lse_ref.detach().float(),
    attn_sink, topk_idxs, softmax_scale=softmax_scale)
(d_sink_ref,) = torch.autograd.grad(out_ref, (sink_ref,), dout.to(torch.float64))
torch.cuda.synchronize()
torch.testing.assert_close(r["d_sink"].double(), d_sink_ref, atol=2e-5, rtol=2e-2)  # FAILS
```

The full sweep was run as black-box regression tests (public wrapper + FP64
reference only, no kernel internals) under
`test/python/fe_api/dsa/test_DSA_sparse_attention_backward.py`. Because the
invalid-index case faults the CUDA context and the empty-row case deadlocks the
GPU, each case must be run in its own process, e.g.:

```bash
cd test/python
pytest -p no:cacheprovider fe_api/dsa/test_DSA_sparse_attention_backward.py \
       -m "L0 or L1" -k "masks_invalid or infinite_sink or empty_row or \
           dsink_uses_recomputed_delta or common_logit_shift or saturated_softmax or \
           high_activation_scale or valid_counts" \
       --forked --timeout=150
```

### Expected behavior

- Invalid / padded / out-of-range / past-`length` top-k entries contribute zero
  probability and zero gradient; `dq`, `dkv`, `d_sink` stay finite.
- A fully-empty row (`topk_length = 0`) produces zero output and zero gradients
  and returns immediately (no hang).
- A saturating sink (`attn_sink → +inf`) yields zero key/query gradient.
- `d_sink` matches the FP64 autograd reference within `atol=2e-5, rtol=2e-2`.
- `dq`/`dkv` match the FP64 reference within 5% relative L2 under any common logit
  shift, saturated softmax, or high activation scale (softmax gradients are
  invariant to a common additive shift of the logits).

### Actual behavior

Run against the unmodified kernel as shipped in `nvidia-cudnn-frontend` 1.27.0
(current `develop`), on one H200 (SM90). SM100-only tests skip automatically.

| Condition | Observed | Expected |
|---|---|---|
| `topk_idxs` row with `-1` padding / OOB / past-`length` entries | `dq` contains **NaN** | 0 probability, 0 gradient |
| `topk_length = 0`, sink disabled | **GPU deadlock** (synchronize never returns) | 0 grad, immediate return |
| `attn_sink = +inf` (sink owns all mass) | `dq` **all NaN** | 0 key/query gradient |
| `d_sink` vs FP64 reference | **128/128 heads wrong**, max abs 9.77e-4 | within atol 2e-5 |
| Common logit shift ∈ {−128, −90, −80, +80} | `dq` **NaN** | unchanged from zero-shift result |
| Saturated softmax (dominant gap ≥ 5 nats) | `relative_l2` 15.6% … 100%+ | ≤ 5% |
| High activation scale ∈ [8, 32] | `relative_l2` up to 16800%, or NaN | ≤ 5% |
| Random valid counts 1–127, scale ∈ {16, 28} | `dq` **NaN** | ≤ 5% |

**Test-suite result: 8 PASS, 34 FAIL, 1 HANG, out of 43 SM90-runnable cases.**

| Test (SM90) | Level | Result |
|---|---|---|
| `masks_invalid_topk_rows[negative-index-holes]` | L0 | **FAIL** — `dq` NaN |
| `masks_invalid_topk_rows[partial-topk-tile]` | L0 | **FAIL** — `dq` NaN |
| `masks_invalid_topk_rows[short-length]` | L0 | **FAIL** — `dq` NaN |
| `masks_invalid_topk_rows[active-negative-index]` | L0 | **FAIL** — `dq` NaN |
| `masks_invalid_topk_rows[active-out-of-range-index]` | L0 | **FAIL** — `dq` NaN (CUDA fault) |
| `masks_invalid_topk_rows[length-exceeds-index-width]` | L0 | **FAIL** — `dq` NaN |
| `empty_row_without_sink_is_finite` | L0 | **HANG** — synchronize never returns (>150 s) |
| `infinite_sink_is_finite` | L0 | **FAIL** — `dq` all NaN |
| `dsink_uses_recomputed_delta[90]` | L0 | **FAIL** — 128/128 heads wrong, max abs 9.77e-4 |
| `finite_sink_dq_dkv_do_not_depend_on_requesting_d_sink` | L0 | PASS |
| `fp64_sparse_attention_oracle_ignores_unselected_...` | L0 | PASS (reference self-check) |
| `finite_sink_preserves_large_common_k_gradient[90]` | L0 | PASS |
| `equal_dominant_keys[90]` | L0 | PASS |
| `is_stable_under_common_logit_shift[-128 / -90 / -80 / +80]` | L1 | **FAIL** — `dq` NaN |
| `is_stable_under_common_logit_shift[0.0]` | L1 | PASS |
| `is_accurate_for_saturated_softmax[5 / 10 / 20 / 40 / 80 / 120]` | L1 | **FAIL** — up to 100%+ rel-L2 |
| `is_accurate_at_high_activation_scale[8 … 32]` (10) | L1 | **FAIL** — up to 16800% rel-L2 |
| `valid_counts_1_to_127_across_seeds[4.0, 3 seeds]` | L1 | PASS |
| `valid_counts_1_to_127_across_seeds[16.0, 3 seeds]` | L1 | **FAIL** — NaN |
| `valid_counts_1_to_127_across_seeds[28.0, 3 seeds]` | L1 | **FAIL** — NaN |

**Root cause.** Line numbers refer to `dsa_bwd_sm90.py` as shipped in 1.27.0.

- *Symptoms 1, 2 — no validity mask on the top-k index.* `_copy_row` loads the KV
  row from the raw index without a range check, and the score is never masked
  before the softmax:

  ```python
  # dsa_bwd_sm90.py:1305
  token_idx = mTopkIdxs_cur[global_topk_row]
  gKV_row   = mKV_cur[token_idx, None]        # token_idx may be -1, OOB, or a sentinel
  # dsa_bwd_sm90.py:1485
  acc_S_mn[r, c] = cute.math.exp2(acc_S_mn[r, c] * softmax_scale_log2 - tLSErLSE[r], fastmath=True)
  ```

  A large garbage score overflows to `inf` and poisons `dq`. When every index in
  a row is invalid (`topk_length = 0`), the empty-row path leaves the CTA short of
  the barrier the launch waits on → deadlock.

- *Symptoms 3, 4 — sink probability and delta.* The delta is computed as `O·dO`
  from the BF16 output, and the sink log-sum-exp is unguarded for `+inf`:

  ```python
  # dsa_bwd_sm90.py:317
  dpsum = (tOrO.load().to(Float32) * tOrdO.load().to(Float32)).reduce(...)   # delta from BF16 O
  # dsa_bwd_sm90.py:339-341
  sum_exp2 = exp2(lse_log2 - lse_max_log2) + exp2(sink_log2 - lse_max_log2)
  lse_with_sink_log2 = lse_max_log2 + log2(sum_exp2)
  p_sink = exp2(sink_log2 - lse_with_sink_log2)      # +inf - +inf = NaN when sink = +inf
  ```

  With `attn_sink = +inf`, `p_sink = exp2(NaN) = NaN`, and `-p_sink * dP_sum`
  writes NaN into `d_sink` / `dq` (symptom 3). Independently, the BF16-rounded
  `O·dO` delta loses the term that only cancels in full precision (symptom 4).

- *Symptom 5 — single-pass online statistics.* `Σ p·dP` is accumulated in one pass
  relative to a block-local maximum; when the global maximum differs from the
  running estimate at the reduction point, the reduction cancels catastrophically.

**Suggested fix (direction only).**

1. Mask top-k lanes whose index is `< 0`, `>= s_kv`, or beyond `topk_length`
   before the softmax, and take an explicit zero-gradient / early-return fast path
   for fully-empty rows so every CTA still reaches the barrier.
2. Guard the sink log-sum-exp against `+inf − +inf` and the empty-row `lse = -inf`
   case, so no `NaN` / `0·inf` reaches `d_sink` or `dq`.
3. Recompute the softmax delta from `Q·K·dO` in full precision for the sink /
   `d_sink` path instead of reusing the BF16 `O·dO` product.
4. Make the `Σ p·dP` statistics numerically stable (exact online dominant/tail
   accounting, or a two-pass reduction against the true global maximum).

### Environment (where applicable)

```
cuDNN Frontend Version: 1.27.0 (develop)
cuDNN Backend Version: 9.2.0 (backend_version() == 92000)
CUDA Toolkit/Runtime Version: 13.2
GPU: NVIDIA H200 (compute capability 9.0, SM90)
CUDA driver: 595.58.03
OS: Linux 6.8.0-124-generic
Python: 3.10
Framework: PyTorch 2.13.0+cu130
nvidia-cutlass-dsl: 4.7.0
```

SM100 (Blackwell) was not exercised because this host reports SM90; the SM100
backward path shares the same delta / sink / statistics structure and is expected
to show the same symptoms.

### Last known working version

commit 651a732b

**Comment by wangfanChris** [comment]
The SM100 kernel already guards parts of cases via #439 (zero-length top-k rows) and #421 (sink normalization), but the SM90 path in dsa_bwd_sm90.py carries no equivalent guards, so the empty-row hang and infinite-sink NaN still reproduce on Hopper. The remaining symptoms (unmasked invalid top-k → NaN dq, BF16 O·dO d_sink, and the single-pass online-softmax accuracy collapse) affect both paths and are not addressed by any merged PR.

**Comment by Anerudhan** [comment]
Thanks for the issue. 
Will ask the team to triage.

**Comment by SuperGoodGame** [comment]
I've been looking at this on an H200 (cc 9.0, so only the SM90 path runs here) and can reproduce 1, 2 and 3.

For 2, I bisected the hang. Wrapping only the `sdS_consumed` wait (`dsa_bwd_sm90.py:1267`) in `if topK > 0` still hangs — it also needs the `G4_half_ready` wait at `:1545`. Both are 256-thread `bar.sync`es, and with `topk_length == 0` WG1 runs zero mainloop iterations and exits, so WG0 waits on arrivals that never arrive. I'd rather not guard the individual barriers; I'd like to add an early exit before any barrier or pipeline init, same shape as #439 on SM100. That needs the raw gmem dQ passed into the kernel, since the TMA views can't serve the zero-store.

For 3, there are two `+inf - +inf` sites when the sink saturates: the LSE-with-sink at `:386-394` and the `d_sink` block at `:333-348`. `fmax` picks `+inf`, then `exp2(inf - inf)` is NaN. The existing guard only covers `lse == inf`, not the sink. SM100 has the same expression, but it auto-skips on this host so I can't test it — happy to leave that to you.

One contract question before I write 1: is the caller required to pass `-1` for padded slots, and are out-of-range positive indices invalid input the kernel must tolerate? `dsa.md` documents `-1` for the indexer's output, but the backward section doesn't say. The kernel's own `token_idx >= 0` check at `:1448` suggests `-1` is expected.

Could you assign 1, 2 and 3 to me? I'd start with 2 on its own so you can see the shape before the rest. Leaving 4 and 5 alone: 4 is a cost/accuracy trade-off that's yours to pick, and for 5 I can't reproduce the +80-logit-shift case — the file has no online softmax and no running max, so I don't think that's the mechanism.

**Comment by Anerudhan** [comment]
Thanks @SuperGoodGame assigned this  to you for now.
